### ISSUE TITLE
Why is the same agent brain instance used between worlds from the same builder?

### ISSUE THREAD

**Issue by jwaa** [matrx-software/matrx#262]
**What is your question?**
Any `AgentBrain` in a world from the same world builder is the same instance. I expected that every world that I run from the same world builder, would use a different and new instance of each of the agent's brains. Why is this not the case? What does that mean for me? And do I need to do some resetting for myself?

**To what is your question related?**
- [ ] I am not sure if I should use MATRX.
- [x] I have a question about the idea behind MATRX.
- [ ] I don't know how to do something with MATRX.
- [ ] I don't know where to find something.
- [ ] I think I found a bug, but I am not sure.
- [ ] I want to contribute, but don't know where to start.
- [ ] Something else.

**Comment by jwaa** [comment]
This is related to the idea of why we seperated the agent's brain (e.g. the code that decides actions) and its body (e.g. its digital embodiement that interacts or is interacted upon in the world). We wanted MATRX to support learning agents; agents whose brain learn over time by interacting with different and multiple worlds. Hence, we decided to have this seperation in place. As it allowed us to create a new agent body for every new world but with the same brain instance.

If you instantiate an `AgentBrain`, provide it to a `WorldBuilder` when adding a new agent, create a new world and run it, the next time you create another world from that same `WorldBuilder` that agent will have a new body but its brain will be capable of remembering what happened in the previous world.

To support this functionality we added a method `WorldBuilder.initialize()` which is called by every newly created `GridWorld` to allow the user to reset anything that should be reset between world runs (or to add new things). 

Overall this functionality is not clear documented, specifically also not in `AgentBrain.initialize()`. There is a current issue active (#264) to fix this. 

Note that there is a currently open bug (#263) where MATRX does not reset public `AgentBrain` attributes such as `AgentBrain.previous_action`. It is left to you to do that, which is quite inconvenient for you. So we will try to fix that as soon as possible.
